paasta_tools/marathon_tools.py

```diff
@@ -204,33 +203,41 @@ def get_instances(self):
         :returns: The number of instances specified in the config, 0 if not
                   specified or if desired_state is not 'start'."""
         if self.get_desired_state() == 'start':
-            instances = self.get_min_instances()
             max_instances = self.get_max_instances()
             if max_instances is not None:
+                instances = self.get_min_instances()
                 try:
                     zk_instances = get_instances_from_zookeeper(
                         service=self.service,
                         instance=self.instance,
                     )
                 except NoNodeError:  # zookeeper doesn't have instance data for this app
-                    pass
+                    return instances
                 else:
-                    instances = max(
+                    return max(
                         min(zk_instances, max_instances),
                         instances,
                     )
-            return instances
+            else:
+                return self.config_dict.get('instances', 1)
         else:
             return 0
 
-    def get_backoff_seconds(self):
+    def get_autoscaling_params(self):
+        default_params = {
+            'method': 'default',
+        }
+        return self.config_dict.get('autoscaling', default_params)
+
+    def get_backoff_seconds(self, instances=None):
         """backoff_seconds represents how many seconds a penalization factor for failing tasks.
         Every time a task fails, Marathon adds this value multiplied by a backoff_factor.
         In PaaSTA we know how many instances a service has, so we adjust the backoff_seconds
         to account for this, which prevents services with large number of instances from
         being penalized more than services with small instance counts. (for example, a service
         with 30 instances will get backed off 10 times faster than a service with 3 instances)."""
-        instances = self.get_instances()
+        if instances is None:
+            instances = self.get_instances()
```

Comment:
I understand why you are doing this, but another pattern you should consider is the "memoize" pattern if you haven't heard of it.

I think in paasta this pattern should be safe considering how we run things